### PR TITLE
[Qt] Payment request expiration bug fix (re-done)

### PR DESCRIPTION
**Pull request by Diapolo**
- this is based on #4122 (which can be closed)

Currently a payment request is only checked for expiration upon receipt.
It should be checked again immediately before sending coins to prevent
the user from paying to an expired invoice which would then require a
customer service interaction.
- add static verifyExpired() function to PaymentServer to be able to use
  the same validation code in GUI and unit-testing code
- extend unit tests to use that function and also add an unit test which
  overflows, because payment requests allow expires as uint64, whereas we
  use int64_t for verification of expired payment requests

### CODE REVIEW

**Comment by jonasschnelli** [comment]
@Diapolo would it be possible to cover this with a test in `paymentservertests.cpp`?

**Comment by Diapolo** [comment]
One could craft and add an expired payment request into ``paymentrequestdata.h``, copy and paste the check used in this pull ``(details.has_expires() && (int64_t)details.expires() < GetTime()))`` and use that in a testcase, yes. I can't craft a usable payment request usable there, feel free to jump in :).

**Comment by Diapolo** [comment]
Btw. @laanwj ``expires`` is also an uint64 casted to int64_t, is that safe? ``GetTime()``is using int64_t as return value, wheras the used ``time()`` in it is returning ``time_t`` (which is a long long on my compiler suite).

**Comment by gavinandresen** [comment]
Definitely needs a unit test (or two, testing expires with large unsigned values is a great idea).

@Diapolo : if you aren't willing to create unit tests for your code, then, in my humble opinion, you shouldn't be writing code. Testing is critical.

**Comment by Diapolo** [comment]
@gavinandresen Sorry if you have the impression I'm not willing, it's more my skills are limited in that case. I really have no idea how to craft a suiting payment request for ``paymentrequestdata.h``, like you did -_-.

**Comment by jonasschnelli** [comment]
@Diapolo i'll try to create one a link you in a commit for `paymentrequestdata.h`

**Comment by Diapolo** [comment]
@jonasschnelli I'm glad if you also can help me in understanding how you do it :).

**Comment by gavinandresen** [comment]
There's simple c++ code for creating payment requests here:  https://github.com/gavinandresen/paymentrequest/blob/master/c%2B%2B/paymentrequest-create.cpp

I used it to generate a binary payment request (after using openssl commands to generate private keys, certificates, etc).

Then I converted the binary payment request files to base64 (e.g. openssl enc -base64)

Then copy the base64 into the .h file.

**Comment by Diapolo** [comment]
@gavinandresen Thanks for linking that in here, I'll take a look.

**Comment by jonasschnelli** [comment]
Could you not take one from https://bitcoincore.org/~gavin/createpaymentrequest.php and base64 it?
If you need different values, take the php source, run them on MAMP/XAMP, etc. 
I suppose this might also work.

**Comment by jonasschnelli** [comment]
@Diapolo ping me if you didn't succeed then i start looking at it.

**Comment by Diapolo** [comment]
@gavinandresen Wow, that was quite hard up to where I am now :). I took your whole branch https://github.com/gavinandresen/paymentrequest and converted the entire create_ca.sh to a Windows compatible create_ca.cmd, reworked build_detect_platform, Makefile, paymentrequest-create.cpp, paymentrequest-dump.cpp and even the README.txt (to perhaps aid others with this stuff) to be Windows compatible!

I build the executables with MinGW 4.9.2 x64 and the ca stuff with OpenSSL 1.0.1j.

Now when I try to do ``paymentrequest-create paytoaddress=1BTCorgHwCg6u2YSAWKgS17qUad6kHmtQW memo="Just Testing" amount=11.0 | paymentrequest-dump``, I get the following error, which I could not solve until now:
<pre>
[libprotobuf ERROR google/protobuf/message_lite.cc:123] Can't parse message of type "payments.PaymentRequest" because it is missing required fields: serialized_payment_details
</pre>

Can you or any other (@mhearn) assist with solving this!?

**Comment by Diapolo** [comment]
I depend on #5642 getting merged, before I can add another unit test for this.

**Comment by jonasschnelli** [comment]
@Diapolo could you not include #5642 commits in this PR and add a UT?

**Comment by Diapolo** [comment]
@jonasschnelli Yeah I could rebase this ontop of the other pull, but I'm not sure this allows for a quicker merge. The other pr requires a review of at least one other core dev I guess...

**Comment by Diapolo** [comment]
Rebased to be on top of #5642, which has 2 ACKs now.

**Comment by Diapolo** [comment]
If this passes Travis it should be ready to merge. Needs review by @gavinandresen or @laanwj.

**Comment by Diapolo** [comment]
So 8 days ago I finished this medium priority pull and no one cares to review/ACK? Are we dropping wallet support or why is no one caring about our reference implementation of the payment protocol?

**Comment by jonasschnelli** [comment]
I think we all care.
It's just very hard to set priorities.
Your work looks really good and I will do a proper test and a review.

Please be patient and give us some more days. Thanks.

**Comment by jonasschnelli** [comment]
Here are some fresh built binaries from my gitian-build-server (in case some likes to test but not to build):
https://bitcoin.jonasschnelli.ch/pulls/5620/

**Comment by jonasschnelli** [comment]
Tested, reviewed
ACK.

bildschirmfoto 2015-01-29 um 09 24 47

File: src/qt/paymentserver.cpp [hunk 1]
```diff
@@ -756,3 +753,15 @@ bool PaymentServer::verifyNetwork(const payments::PaymentDetails& requestDetails
     }
     return fVerified;
 }
+
+bool PaymentServer::verifyExpired(const payments::PaymentDetails& requestDetails)
+{
+    bool fVerified = (requestDetails.has_expires() && (int64_t)requestDetails.expires() < GetTime());
```
Comment (SergioDemianLerner) [inline src/qt/paymentserver.cpp]: Shouldn't "fVerified" be renamed "fExpired" ? "Verified" implies something positive.

**Pull request merged**